minor-modes.lisp

```diff
@@ -0,0 +1,1113 @@
+                                   (setf curmap topmap))
+                                 keymap)
+               (abort-bindings ()
+                 :report "Return the keymap without binding further keys"
+                 topmap)
+               (abort-bindings* ()
+                 :report "Return an empty keymap"
+                 (stumpwm:make-sparse-keymap)))
+             topmap)
+            (t (restart-case
+                   (error "Function MAKE-MINOR-MODE-KEYMAP cant understand ~A"
+                          keymap)
+                 (use-empty-keymap ()
+                   :report "Use an empty keymap"
+                   (stumpwm:make-sparse-keymap))))))))
+
+(defun make-minor-mode-keymap (spec)
+  (generate-keymap spec))
+
+(defun make-minor-mode-top-map (top-map-spec root-map-spec)
+  "Create a top map for a minor mode based upon its TOP-MAP-SPEC and
+ROOT-MAP-SPEC."
+  (let ((top-map nil)
+        (root-map (if root-map-spec
+                      (make-minor-mode-keymap root-map-spec)
+                      (make-sparse-keymap))))
+    (fill-keymap top-map *escape-key* root-map)
+    (generate-keymap top-map-spec top-map)))
+
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defun make-special-variable-name (mode name)
+    (intern (format nil "*~A-~A*" mode name)))
+  
+  (defun parse-minor-mode-options (options)
+    (let ((valid-options
+            '((:interactive            . 1)
+              (:scope                  . 1)
+              (:global                 . 1)
+              (:lighter-make-clickable . 1)
+              (:lighter                . 1)
+              (:expose-keymaps         . 1)
+              (:rebind                 . 1)
+              (:root-map               . 1)
+              (:top-map                . 1)
+              (:enable-when            . t)
+              (:make-hooks             . 1)
+              (:default-initargs       . t)
+              (:define-command-definer . 1)))
+          (all-vals '())
+          (other-opts '()))
+      (flet ((collect-values (option)
+               (destructuring-bind (option-name &rest option-arguments) option
+                 (let ((argcount (cdr (assoc option-name valid-options))))
+                   (if argcount
+                       (progn (if (and (numberp argcount)
+                                       (= argcount 1))
+                                  (push (car option-arguments) all-vals)
+                                  (push option-arguments all-vals))
+                              (push option-name all-vals))
+                       (push option other-opts))))))
+        (mapc #'collect-values options)
+        (values all-vals other-opts))))
+
+  (defun define-command-macro (mode)
+    `(defmacro ,(intern (string-upcase (format nil "define-~A-command" mode)))
+         (name (&rest args) (&rest interactive-args) &body body)
+       (multiple-value-bind (bod decls docstring)
```

Comment:
what do you think about just shadowing body instead of using bod because body is already in scope?
```suggestion
       (multiple-value-bind (body decls docstring)
```

Reply:
I think its a good move, and I implemented it.